From cc870dd48fd59832789bd5971c8a922895c81924 Mon Sep 17 00:00:00 2001
From: Maciej Szymanski <mszymans@qti.qualcomm.com>
Date: Sun, 17 Aug 2025 16:07:05 +0200
Subject: [PATCH] vrend define missing gbm formats

Signed-off-by: Maciej Szymanski <mszymans@qti.qualcomm.com>
---
 src/vrend_winsys_gbm.h | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/src/vrend_winsys_gbm.h b/src/vrend_winsys_gbm.h
index cd92e0f2..b0ae5f2b 100644
--- a/src/vrend_winsys_gbm.h
+++ b/src/vrend_winsys_gbm.h
@@ -35,6 +35,16 @@
 #define VIRGL_GBM_MAX_PLANES 4
 #endif
 
+/* GBM_FORMAT_R8 */
+#ifndef GBM_FORMAT_R8
+#define GBM_FORMAT_R8		__gbm_fourcc_code('R', '8', ' ', ' ') /* [7:0] R */
+#endif
+
+/* GBM_FORMAT_XBGR16161616F */
+#ifndef GBM_FORMAT_XBGR16161616F
+#define GBM_FORMAT_XBGR16161616F __gbm_fourcc_code('X', 'B', '4', 'H') /* [63:0] x:B:G:R 16:16:16:16 little endian */
+#endif
+
 /* GBM_FORMAT_ABGR16161616F was added since mesa 20.0 */
 #ifndef GBM_FORMAT_ABGR16161616F
 #define GBM_FORMAT_ABGR16161616F __gbm_fourcc_code('A', 'B', '4', 'H') /* [63:0] A:B:G:R 16:16:16:16 little endian */
-- 
2.25.1

